internal/terraform/context_validate_test.go

```diff
@@ -2031,3 +2031,50 @@ resource "test_object" "t" {
 		t.Fatal(diags.ErrWithWarnings())
 	}
 }
+
+func TestContext2Plan_lookupMismatchedObjectTypes(t *testing.T) {
+	p := new(MockProvider)
```

Comment:
Given that we were all so surprised that this had ever worked before, I think it would help to include a comment here explaining _why_ the behavior described by this test is "correct", because I expect that in the far future we'll end up back here after seeing this test fail and have lost the context for what exactly this is testing, and thus might end up incorrectly concluding that the test is wrong.

Reply:
Thanks, I meant to add that and will include it before merging!